### PR TITLE
Major API changes

### LINKED ISSUE #48: API review
@mikeal and I want to review the current API and see how they can be made more ergonomic.

The goal is to have a nice to use and easy to implement API which aligns well with the current vision of IPLD. Breaking the API is possible and even very likely.

Currently also the spec and the format implementations are not in sync at the moment (see the open issues for more information).

Parts of it is that it will use async/await, hence aligns well with the [Awesome Endeavour: Async Iterators](https://github.com/ipfs/js-ipfs/issues/1670).

We can use this issue for further discussion/referencing things.

/cc @hacdias

### PR DESCRIPTION
**Pull request by vmx**
The whole IPLD APIs get a review to make them more consistent and
easier to use.

The biggest change is that there's not `resolve` API anymore. From
now on you access the properties of the JavaScript objects directly.

Issues that were taken into account:

 - https://github.com/ipld/interface-ipld-format/issues/31
   - [x] Remove `isLink()` method from formats: `isLink()` is no
         longer needed as all links will be CID objects you can easily
         check for
 - https://github.com/ipld/interface-ipld-format/issues/44
   - [x] Proposal: Move resolver to use CID instances for links: Not
         applicable anymore as `resolve.resolve()` is removed
 - https://github.com/ipld/interface-ipld-format/issues/46
   - [x] properties: align spec with implementation: Covered by spec
 - https://github.com/ipld/interface-ipld-format/issues/49
   - [x] Define how `toJSON()` should look like: Binary and CIDs are
         defined with example
 - https://github.com/ipld/interface-ipld-format/issues/34
   - [x] Implementation of nested objects: Won’t fix as `tree()` is
         not part of the API anymore

Closes #48.

### CODE REVIEW

File: README.md
Comment (mikeal) [inline README.md]: So, we cut the "canonical serialization" stuff from IPLD a while back. That means this method probably shouldn't be required and that any conversions from one format to another are custom.

Given that we don't have a method that takes this representation and turns it back into a serialized block should we include this?

Comment (vmx) [reply README.md]: It's for third party applications that want to interact with IPLD stuff. What I have in mind is [search on IPLD](https://github.com/ipld/ipld/issues/43). So having that built-in would be super nice. Else I'd probably end up monkey patching all formats, which is also OK.

Comment (mikeal) [reply README.md]: The thing is, we've sort of opened the door to formats having types that they can't easily, or potentially don't want to, convert to a JSON representation.

Even the base types we require for the data model are *base* requirements, there's nothing keeping a format from having more types and there's no guarantee you could represent those in JSON.

Rather than put this on the codec's to write I would just write something for "search on IPLD" that converts an object to a JSON representation and is aware of Binary and CID types. That will work with anything that supports the data model and you could just pass it the deserialized node.

Comment (vmx) [reply README.md]: Very good points. I was just so keen to have a JSON representation. But it can indeed live outside of IPLD Format API which should really be kept to a minimum. I'll remove that part.

**Comment by vmx** [comment]
@mikeal I've removed `toJSON()`. I'd like your feedback especially on using the multicodec codecs byte values and not their string identifiers. I think we should do that in all of our APIs. What do you think?

File: README.md
Comment (mikeal) [inline README.md]: If we are moving to using the hex value rather than the string value do we actually need to require in *this* API that the code is already registered in the table?

At some point this table is going to get much harder to add things to and it would be nice to open up an easier path for developers who are working on new codecs.

Comment (mikeal) [reply README.md]: To be clear, I'm recommending that we change from "It needs to" to "It *should* be".

**Comment by mikeal** [comment]
@vmx I like that change, I think it has some other implications I've noted inline.

File: README.md
Comment (vmx) [reply README.md]: You're absolutely right. This is what I wanted to say, it's just a language thing. I'll try to formulate it in a clearer way.

**Comment by vmx** [comment]
@mikeal I also changed the hash codec to "should".

Once this change is approved, I'd like to merge it myself as I'd like to get a single clean commit and I need to edit the commit message a bit :)

**Review by mikeal — APPROVED** [review]

**Comment by mikeal** [comment]
@vmx approved :)

**Comment by vmx** [comment]
@daviddias As this is a major change I'd like to get your approval before I merge it too.

File: README.md
Comment (daviddias) [inline README.md:23]: Where did these methods go?

Comment (daviddias) [reply README.md:23]: I would like to see the new `js-ipld` with the formats updated and resolving through things with all of these updates. It seems that we are preemptively removing things but I might be missing something.

Comment (vmx) [reply README.md:23]: From the commit message:

> The biggest change is that there's not `resolve` API anymore. From
> now on you access the properties of the JavaScript objects directly.

So instead of `resolve.resolve(binaryBlob, '/deeply/nested/path)' you deserialize  the node first (which resolve also needs to do) and then on the resolved node, let's cal it `ipldNode` you do `ipldNode.deeply.nested.path`.

For `tree()`-like functionality (if needed) you do what [dag-json is doing](https://github.com/mikeal/dag-json/blob/3ba08ee1ac92341038585d07f482dab7caf621e5/index.js#L109-L125).

Comment (daviddias) [reply README.md:23]: So now it is the user that needs to convert an IPLD Path into JS notation?

Comment (daviddias) [reply README.md:23]: Note that dag-pb supported resolve as in `cid/name-of-the-link-to-the-other-node` and `cid/links/0/hash/`. It was basically packing a transformation inside the format. It will be an extra breaking change.

Comment (daviddias) [reply README.md:23]: This looks like a good proposal and I do not intend to push back on it, however,I would still like to see first how this will impact js-ipld internals for the multiple formats we support today.

Comment (vmx) [reply README.md:23]: > So now it is the user that needs to convert an IPLD Path into JS notation?

The user normally would use js-ipld. There you will still have the paths. If users use an IPLD Format directly they will use it as a normal JS obect.

> Note that dag-pb supported resolve as in `cid/name-of-the-link-to-the-other-node` and `cid/links/0/hash/`. It was basically packing a transformation inside the format.

Thanks for bringing this up. `dag-pb` is certainly a special case. It has IPFS and IPLD style traversals. I'll make sure that IPFS style traversals will still be possible. Generally I think IPLD should only support IPLD style traversals, but of course we need to support the `dag-pb` legacy somehow.

Comment (mikeal) [reply README.md:23]: One benefit of not implementing path resolution in the codec interface is that it actually **forces** parity between the in-memory object representation and the path representation. 

In the old interface, because it was a different method, it was possible (and I think this actually happened) where the path resolution was different from the de-serialized object representation.

Comment (vmx) [reply README.md:23]: @daviddias I implemented this spec for dag-pb, also the named links. See https://github.com/ipld/js-ipld-dag-pb/pull/105 and especially https://github.com/ipld/js-ipld-dag-pb/pull/105/commits/f98f09d0e3d1a51aa4c367ecbcb035d822385881 to see how named links would work then.

**Comment by vmx** [comment]
After implementing this spec for dag-pb I learned about enumerable properties. Hence I'd like to add this paragraph to the `deserialize()` section:

> All enumerable properties (the ones that are returned by a `Object.keys()` call) of the deserialized object are considered for resolving IPLD Paths. They must only return values whose type is one of the [IPLD Data Model](https://github.com/ipld/specs/blob/master/IPLD-Data-Model-v1.md).

What do others think about this? /cc @mikeal @daviddias

**Comment by vmx** [comment]
After implementing this spec for dag-pb I learned about enumerable properties. Hence I'd like to add this paragraph to the `deserialize()` section:

> All enumerable properties (the ones that are returned by a `Object.keys()` call) of the deserialized object are considered for resolving IPLD Paths. They must only return values whose type is one of the [IPLD Data Model](https://github.com/ipld/specs/blob/master/IPLD-Data-Model-v1.md).

What do others think about this? /cc @mikeal @daviddias

**Comment by achingbrain** [comment]
Another thing to consider - the performance analysis done by @mcollina and the NearForm team has revealed we spend a lot of time waiting for [artificially induced asynchronous behaviour](https://github.com/ipld/js-ipld-dag-pb/blob/master/src/util.js#L87) to resolve (basically `setImmediate` and friends).

If the serialization/deserialization can be done synchronously we should consider making these synchronous methods instead of taking callbacks or returning promises.

The alternative is to [invoke callbacks synchronously](https://github.com/ipld/js-ipld-dag-pb/blob/master/src/util.js#L75) or use `process.nextTick` but we run the risk of [releasing Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony) - better to just examine whether async is necessary at all in this case.

**Comment by mikeal** [comment]
FYI, I wrote a synchronous version of `dag-cbor` in JS. All of our sync code in that module is self-induced :) https://github.com/mikeal/dag-cbor-sync

**Comment by vmx** [comment]
> If the serialization/deserialization can be done synchronously we should consider making these synchronous methods instead of taking callbacks or returning promises.

Are already resolved promises a big bottleneck? Promises seem like a good trade-off to make async stuff possible.

**Comment by mcollina** [comment]
> Are already resolved promises a big bottleneck? Promises seem like a good trade-off to make async stuff possible.

That will cause the value to be deferred in the microtask queue. In practice, the impact will be similar to `process.nextTick`. It is _better_ than a full `setImmediate` as that is triggered after a full run of the event loop.

Note that `await Promise.resolve(42)` causes multiple tasks (4 in Node v10) to be scheduled in the microtask queue. See https://www.youtube.com/watch?v=DFP5DKDQfOc for more details.

File: README.md
Comment (achingbrain) [reply README.md:23]: A few things about that PR:

In that change named links aren't added to the object if they match existing property names (`size`, `data`, etc) so not all property traversals will work.

If you're doing a traversal in the style of `resolver.resolve(binaryBlob, '/deeply/nested/path')` you have the option of throwing away everything you've traversed so far, whereas if you do `ipldNode.deeply.nested.path` you have to create the whole object in memory first which might not be great depending on your dataset.

Comment (vmx) [reply README.md:23]: > In that change named links aren't added to the object if they match existing property names (`size`, `data`, etc) so not all property traversals will work.

That's the same behaviour as the current implementation, except that now also the lower case versions are "reserved".

> you have the option of throwing away everything you've traversed so far,

You could implement the same semantics with JavaScript Proxies if needed.

Comment (achingbrain) [reply README.md:23]: > That's the same behaviour as the current implementation

Hmm, could we make it better?

> You could implement the same semantics with JavaScript Proxies if needed.

What's the browser support for that like?  I found [babel-plugin-proxy](https://www.npmjs.com/package/babel-plugin-proxy) but the README says it's got really bad performance and don't recommend using it in production environments.

Comment (vmx) [reply README.md:23]: > Hmm, could we make it better?

I would get rid of the named links traversal (I call it "IPFS traversal") or put it under a field (e.g. `ipfs`) so it won't collide. Though I think being just as broken as we are currently is good enough for now, especially since this is again dag-pb specific and shouldn't be a problem with other formats.

> What's the browser support for that like?

Pretty good: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Browser_compatibility

Comment (achingbrain) [reply README.md:23]: > I would get rid of the named links traversal 

This is not a crazy idea.  It's only of limited use for IPFS because there isn't a one-to-one mapping between IPFS paths and IPLD paths for HAMT shards.

E.g.

```javascript
resolve.resolve(binaryBlob, '/deeply/nested/path.txt')
```

Might end up needing to be

```javascript
resolve.resolve(binaryBlob, '/deeply/A0/01/FC-nested/path.txt')
```

..and you don't know what the prefix for each segment is going to be until you descend into that level of the trie as it's dependent on the siblings.

**Comment by daviddias** [comment]
@vmx following your [request for the review of this proposal](https://github.com/ipfs/team-mgmt/pull/795/files). 

This is such a massive change that I don't believe it is wise to merge this one without having sibling PRs to all of its dependents in IPFS land. Can you take the ownership as a builder of IPLD to test the assumptions made by this PR through creating PRs to all the IPFS modules that will require changes. This will give you a complete understanding of how this will impact its usage and also get us ready to merge everything in one swipe.

With these PRs, the Lead Maintainers of which module will also be able to observe how this affects the rest of the code and give direct feedback accordingly. 

Create an [`awesome-endeavour`](https://github.com/ipfs/js-ipfs/issues?q=is%3Aopen+is%3Aissue+label%3A%22awesome+endeavour%22) to track this work.

File: README.md
Comment (vmx) [reply README.md:23]: I mostly kept the named links to show that such things are still possible.

**Comment by vmx** [comment]
@daviddias Of course there will still be changes once things get implemented. For me approval means "looks good for me, I don't see any major problems" and I think that's what I got from you now.

**Comment by daviddias** [comment]
@vmx you have my 👍 of "LGTM, but do try it first with all the other repos before merging and creating a massive breaking change" :)

**Review by alanshaw — APPROVED** [review]

**Comment by vmx** [comment]
@Gozala I'd like your feedback on this one :)

File: README.md
Comment (Gozala) [reply README.md:23]: I personally do think that move to JS based traversal has following disadvantages:

1. As @achingbrain  pointed out that implies that that whole node is de-serialized in one step. Sure JS proxies could be utilized but they are very slow and are really difficult to get right as in ensure that they you don't end up creating memory leaks (we used them a lot for cross compartment boundaries in gecko & in my opinion for non trivial cases e.g `process.env` they're not worth it).

   I have being wanted to implement ipld resolver for [flatbuffers](https://google.github.io/flatbuffers/) who's primary advantage over proto-buffs is laziness - only relevant fields can be decoded without instantiating objects for intermediaries. It seems like with this move taking advantage of this would be impossible.

   Proxies only work if you can decode everything synchronously, if that is not the case then you wind up with bunch of promises you have to await on to get to the nested data, which IMO is less convenient than it is to provide a path.

Comment (Gozala) [inline README.md:23]: Large naming changes might be out of the scope here, but if it isn't  I would favor following terminology:

File: README.md
Comment (Gozala) [inline README.md:17]: `encode(node)`

File: README.md
Comment (Gozala) [inline README.md:18]: `decode(node)`

File: README.md
Comment (Gozala) [inline README.md:22]: `codec`

File: README.md
Comment (Gozala) [inline README.md:60]: Maybe `serialize(IpldNode):Uint8Array|Promise<Uint8Array>` ?

File: README.md
Comment (Gozala) [inline README.md:66]: I would really encourage to remove requirement to return Buffer in favor of Uint8Array. This requirement makes browser use case unnecessarily complicated, and in practice isn't really necessary. Other IPFS libraries can do the casting where necessary.

P.S. node Buffer is also a valid Uint8Array so node case will require no change.

Comment (Gozala) [reply README.md:66]: I would love if there was some word about errors as well. Specifically how decoder can return error and how that error propagates.

File: README.md
Comment (Gozala) [inline README.md]: This is more of a question.

From what I can tell implementation of the `cid` is pretty much the same across the board. Is there reason to require implementation of `cid` in first place ? Could this be made optional, so only implementation that need a custom behavior can provide it ?

File: README.md
Comment (Gozala) [inline README.md:73]: I did wrote more about this above but I think I want add one more note here - Requirement for implementation to provide a proxy just to avoid full decoding makes a high bar, in fact I would argue that custom [`DataView`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) class with getters is probably a far better alternative. However that still  implies 5 allocations when accessing something like `/deeply/A0/01/FC-nested/path.txt` which may seem insignificant but in some cases (e.g. representing Virtual DOM changelists in [dominion](https://github.com/gozala/dominion) library) that makes a difference between fast and unusable performance.

File: README.md
Comment (rvagg) [inline README.md]: this is the only remaining instance of "block", is the intent to remove the concept of "block" from this doc and just use "binary blob"?

File: README.md
Comment (rvagg) [reply README.md:73]: It's not a requirement, _"may also be a Proxy object"_, but I'm very negative on Proxies regardless. Aside from performance concerns, I think we've learned from many previous attempts in other languages that things like Proxies should be reserved for metaprogramming and other reflection use-cases rather than general magical data structures. Pitfalls abound.
Perhaps the way forward is to remove the reference to proxies completely and leave it up to each format and allow experimentation.

Alternatively go the container route which is much nicer--I'm not sure `DataView` is the right container, maybe a bit too low-level, something that feels more like a Map might be more accessible. In fact, a whole custom Map-like that you could `.toObject()` to fully instantiate as a non-lazy object representation might provide the best of both worlds?

```js
let data = await format.deserialize(blob)
data.get('foo') // could be lazy
data.has('/deeply/A0/01/FC-nested/path.txt') // could be lazy
data.toObject() // skip laziness and recursively instantiate
// { foo: 'bar', deeply: { A0: { '01': { 'FC-nested': { 'path.txt': UInt8Array<00, 00, 00, 00> } } } } }
```

It would require specification, and even provision of a custom class, maybe just a simple extension of `Map`.

For deserialization where you know the result is cheap and you want it all you could just instantiate it in the same call:

```js
let copy = Object.extend((await format.deserialize(blob)).toObject(), { lastUpdate: new Date() })
```

```
Map {
  get (key/path)
  has (key/path)
  toObject()
  put() // throws? is mutability helpful here? maybe until it starts to hit awkward edges
}
```

Are there possible use-cases where getters might need to be async or is it safe to frontload that in `decode()`?

Comment (Gozala) [reply README.md:73]: I think you may have misunderstood what I meant by `DataView` - just a container tgat wraps underlying buffer and provides getters to access / decode fields on demand `get key() {`. So pretty much what you described as Map except named getters on prototype instead of `get(key)`

Comment (Gozala) [reply README.md:73]: > Perhaps the way forward is to remove the reference to proxies completely and leave it up to each format and allow experimentation.
 
The whole point of IPLD is to provide common interface for traversing linked data no ? If each implementation is free to design one, the common interface is gone & I think that’s a step in a wrong direction - I much rather have common imperfect interface that can be used across arbitrary formats than have to learn each one

File: README.md
Comment (vmx) [reply README.md:66]: Moving from Node.js to Browser based primitives is certainly on my todo list, but out of scope for these API changes. I don't want to do everything at once.

File: README.md
Comment (vmx) [reply README.md]: An IPLD Format has all the information it needs to calculate the CID. The multicodec code as well as the hash. So you can call `cid()` from layers that build on top of IPLD Formats without knowing those details.

File: README.md
Comment (vmx) [reply README.md:73]: > just a container tgat wraps underlying buffer and provides getters to access / decode fields on demand

But this then won't be lazy for nested objects, will it?

Comment (Gozala) [reply README.md:73]: It will be if for nested records you return corresponding  `DataView`

Comment (Gozala) [reply README.md:73]: It will be if for nested records you return corresponding  `DataView`

Comment (vmx) [reply README.md:73]: And that would be faster then a JS Proxy (I haven't really used Proxies so far, I only toyed around with them)?

*Update*: I played a bit around with Proxies, I remembered them to work differently ;) Conclusion: such a DataView would need similar efforts as a Proxy, but should be faster.

Comment (Gozala) [reply README.md:73]: Few more thoughts on this:

- `DataView` only works in cases where shape of data is static and known ahead of time like protocol buffers however it’s not for generic structures like dictionaries or JSON.
- Proxies also imply syncronous decoding otherwise you’ll have to return promise that may resolve to void or fail (makes me even more concerned)
- If you just return proxy / DataView what the point ? It seems like that can be automated by tge library and provide consistent API + reduce surface for error.

  maybe `dag.view(blob, format)` ahould just do that and utilize lower level API to materialize fields on access.

- It might be worth having two separate endpoints one to decode blob fully & the other to do that per fields. That would allow a user choose right API per use case.

Comment (vmx) [reply README.md:73]: A general note on this discussion thread: My intention of these API changes is to make it easier to use. The current api has `resolve()` which is similar to what @Gozala and @rvagg are proposing. If the things that `deserialize()` returns would just be a JavaScript object, you wouldn't need to learn any new APIs, you could just use it as you're used to do and access all properties directly.

But perhaps  is treating IPLD Nodes as objects is just another layer on-top of lower level building blocks.

Comment (rvagg) [reply README.md:73]: I had a brief conversation with @mikeal that touched on this and it's made me wonder if we're dabbling a bit too much in over-engineering without a use-case that warrants it? I don't know enough yet about the formats so I'll defer to you both, but are there any formats that could justify a lazy parsing layer? We've been messing around with optimizing JSON parsing for years now and we've still not shown that lazy parsing provides any benefits, except perhaps in very extreme cases where the data shape is simple (perhaps predictable) and the data size is large.

So my questions are:

1. Do we imagine large IPLD blocks in common usage, or are we settling on a rough maximum of 1 or 2 Mb (that seems to be what I'm seeing, but only for filesystem type data, everything else I've seen is considerably smaller)
2. Are there any formats that are in common use that could take advantage of lazy parsing with any meaningful performance gain?
3. Are there any potential formats that could take advantage of lazy parsing for meaningful performance gain, even with Flatbuffers @Gozala, is the benefit even meaningful under 1 or 2 Mb since you still have some up-front overhead, i.e. lazy is never free?

When I talk about "meaningful" gain, it has to be in comparison to all of the other weak spots, already I'm finding that the local-only disk storage is painfully slow for simple blocks, p2p storage has orders of magnitude more performance problems, I can't imagine that the performance optimizations we're talking about can even touch the gains we're talking about getting out of this in exchange for some combination of implementation complexity or API complexity.

I notice that a Q4 target @ https://github.com/ipld/roadmap/pull/6/files#diff-04c6e90faac2675aa89e2176d2eec7d8R32 is about returning to the interfaces after "practical" use of them. So maybe it's best to keep it simple and figure out where the weak performance spots are and returning later?

Comment (Gozala) [reply README.md:73]: >  I don't know enough yet about the formats so I'll defer to you both, but are there any formats that could justify a lazy parsing layer?

As I mentioned previously https://github.com/google/flatbuffers/ is a great example of such data format. It's main advantage over protocol buffers is exactly due to fact that it does not require allocations to decode relevant fields. Here is also good article that goes in more details about this https://code.fb.com/android/improving-facebook-s-performance-on-android-with-flatbuffers/ and perf comparison with other formats https://google.github.io/flatbuffers/flatbuffers_benchmarks.html 

> Do we imagine large IPLD blocks in common usage, or are we settling on a rough maximum of 1 or 2 Mb (that seems to be what I'm seeing, but only for filesystem type data, everything else I've seen is considerably smaller)

It's actually less about the size and more about the number of fields that block contains. As mentioned decoding DOM changelist in such format and decoding it had a huge difference in practice, not because change-list itself was large blob but because it caused significant number of allocations that put pressure on GC. Switching to flatbuffers made that take almost no time, especially because it was incremental process weaved with dom mutations.

> Are there any formats that are in common use that could take advantage of lazy parsing with any meaningful performance gain?

I don't think there are today. However decision to make such use cases impossible or significantly more difficult should probably require good amount of consideration - what is the benefit of removing `.resolve` ? How does that benefit fare against making lazy use cases second class ?

> Are there any potential formats that could take advantage of lazy parsing for meaningful performance gain, even with Flatbuffers @Gozala, is the benefit even meaningful under 1 or 2 Mb since you still have some up-front overhead, i.e. lazy is never free?

Again it's much less about the size of the buffer and more about:
1. Number of allocations involved. 
2. When does the actual processing happens (e.g. ahead of time, when needed, never).

> When I talk about "meaningful" gain, it has to be in comparison to all of the other weak spots, already I'm finding that the local-only disk storage is painfully slow for simple blocks, p2p storage has orders of magnitude more performance problems, I can't imagine that the performance optimizations we're talking about can even touch the gains we're talking about getting out of this in exchange for some combination of implementation complexity or API complexity.

I think assumption here is that supporting lazy case should come with API complexity. I simply don't think that's true. API could be very simple - additional function, that can be optional which takes path along with a blob. Lazy use case can implement that function and avoid  doing irrelevant work,  as of eager use cases they can choose not to implement it as `ipfs.dag` layer can provide fallback.

Comment (Gozala) [reply README.md:73]: Point that I was mostly trying to make that switching to this new API comes with trade-off, specifically it prevents lazy decoding and requires allocations where former API did not. This can be avoided without new fancier API, while such API still can be provided as another layer on top of more primitive that makes no assumptions about laziness and will likely would make format implementations simpler.

Whether that is worthy trade-off is maintainers call, I personally incline to think it isn't.

Comment (rvagg) [reply README.md:73]: @Gozala so specifically your recommendation on that is:

> It might be worth having two separate endpoints one to decode blob fully & the other to do that per fields. That would allow a user choose right API per use case.

Are you also saying that a DataView-style approach will be helpful in addition or as part of the per-field decode API?

My suggestion would be to optimise for, in rank order: (1) usability for consumer (simple and familiar feeling API), (2) usability for implementer (lower complexity) and (3) performance. At the moment I'm seeing arguments about performance to be mostly hypothetical, which is one of the main causes of over-engineering.

Comment (mikeal) [reply README.md:73]: I’d recommend avoiding the complexity of accounting for Proxy objects in this release. Once we have a working prototype that needs this feature we can update the interface.

This will not be the last iteration of this API :)

Comment (Gozala) [reply README.md:73]: > Are you also saying that a DataView-style approach will be helpful in addition or as part of the per-field decode API?

I don’t have formed opinion on this. All I was trying to piont out that JS object interface could be added as a layer on top of more low-level API that doesn’t imply debated tradeoffs

> My suggestion would be to optimise for, in rank order: (1) usability for consumer (simple and familiar feeling API)

Ok so is following more usable 

```js
let a = await deserialize(blob).nested
let b = await a.b
let c = await b.c
```

Or

```js
let c = await resolve(blob, "/nested/b/c")
```

I’d argue that later

Comment (Gozala) [reply README.md:73]: > (2) usability for implementer (lower complexity)

For implementing eager case this will make no difference. For lazy case avoiding proxies and lazy getters would be significantly simpler than needing to have them

Comment (rvagg) [reply README.md:73]:
```
let a = await deserialize(blob).nested
let b = await a.b
let c = await b.c
```

When I wrote my initial response in this thread I was trying to imagine a use-case where the properties would need to be asynchronously accessed and I couldn't see an obvious one _except_ for very large blobs that need to be stored on disk--but those wouldn't be passed in to a `deserialize(blob)` call in the first place. Maybe there's a case if there was some _really_ heavy crypto applied to sections of a blob, but even then it's questionable. Async computational tasks are usually made pointless by the boundary crossing problems and is why async methods in Node's `'crypto'` module have been mostly resisted, computationally intensive thread offloading can't be demonstrated except for very heavy work. The browser is a slightly different case because you're not maximising core use as with a server and responsiveness is important, maybe that's a worthy argument? I'm doubtful.

So again, it looks like premature optimisation in search of a use-case.

> > (2) usability for implementer (lower complexity)
>
> For implementing eager case this will make no difference. For lazy case avoiding proxies and lazy getters would be significantly simpler than needing to have them

Agreed, and it'll help make quicker progress on this PR!

Comment (Gozala) [reply README.md:73]: I care about JS implementation in browser context (on server go is likely a better choice anyway). In [browser crypto APIs](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) are async (for good reasons) ignoring that seems strange to me.

And honestly it’s not about optimizations, just about not making currently well supported use cases second class.

P.S. I am working on crypto heavy use case for browsers right now and this API channge will require me to switch to handrolled solution.

Comment (rvagg) [reply README.md:73]: @Gozala ok this is good context, excuse my ignorance. In your use case, is it necessary that the crypto be handled at the format layer or is it reasonable to push it up to ipld itself? I was imagining https://github.com/ipld/ipld/issues/64 being something in ipld and the blocks just providing the encrypted binary data. Perhaps an encrypted block coming from a format gets transposed into an unencrypted ipld block where the key is provided but this happens at the ipld layer, not in the format. Or are you seeing a different mechanism? Again, I'm fairly new to thinking about this stuff so additional insight is appreciated!

Comment (vmx) [reply README.md:73]: I've created https://github.com/ipld/interface-ipld-format/pull/51 which hopefully accommodates all the points brought up here. It's a smaller change to the current API. Let's have a discussion over there.

**Comment by vmx** [comment]
I'm closing this one to prevent confusion. The current discussion on an improved API are at https://github.com/ipld/interface-ipld-format/pull/51.

**Pull request closed without merging**
